File: C++/04_control_flow.cpp
// 04 - Control Flow
// Control flow decides WHICH code runs and WHEN.

#include <iostream>

int main() {

    // =========================================================
    // IF / ELSE IF / ELSE
    // =========================================================
    int temperature = 22;

    if (temperature > 30) {
        std::cout << "It's hot outside.\n";
    } else if (temperature > 15) {
        std::cout << "Nice weather.\n";       // <-- this branch runs
    } else {
        std::cout << "It's cold outside.\n";
    }

    // --- Ternary operator: compact if/else for single values ---
    std::string verdict = (temperature > 20) ? "warm" : "cool";
    std::cout << "The weather is " << verdict << ".\n";


    // =========================================================
    // SWITCH
    // Best when comparing one variable against fixed values.
    // =========================================================
    int day = 3; // 1 = Monday, 7 = Sunday

    switch (day) {
        case 1: std::cout << "Monday\n";    break;
        case 2: std::cout << "Tuesday\n";   break;
        case 3: std::cout << "Wednesday\n"; break; // <-- runs
        case 4: std::cout << "Thursday\n";  break;
        case 5: std::cout << "Friday\n";    break;
        default: std::cout << "Weekend!\n"; break; // runs if no case matches
    }
    // "break" stops execution from falling through to the next case.


    // =========================================================
    // WHILE loop — runs as long as condition is true
    // =========================================================
    std::cout << "\nCounting up with while:\n";
    int i = 1;
    while (i <= 5) {
        std::cout << i << " ";
        i++;
    }
    std::cout << "\n";


    // =========================================================
    // DO-WHILE loop — always runs the body at least once
    // =========================================================
    std::cout << "\nDo-while (starts at 10, condition already false):\n";
    int j = 10;
    do {
        std::cout << j << " "; // prints 10 even though j > 5
        j++;
    } while (j <= 5);
    std::cout << "\n";


    // =========================================================
    // FOR loop — best when you know the iteration count
    // =========================================================
    std::cout << "\nCounting down with for:\n";
    for (int k = 5; k >= 1; k--) {
        std::cout << k << " ";
    }
    std::cout << "\n";


    // =========================================================
    // BREAK and CONTINUE
    // =========================================================
    std::cout << "\nBreak at 4:\n";
    for (int n = 1; n <= 10; n++) {
        if (n == 4) break;        // exit the loop entirely
        std::cout << n << " ";
    }
    std::cout << "\n";

    std::cout << "\nSkip even numbers with continue:\n";
    for (int n = 1; n <= 10; n++) {
        if (n % 2 == 0) continue; // skip to next iteration
        std::cout << n << " ";
    }
    std::cout << "\n";

    return 0;
}

File: JavaScript/04_control_flow.js
// 04 - Control Flow

// =========================================================
// IF / ELSE IF / ELSE
// =========================================================

const score = 75;

if (score >= 90) {
    console.log("A");
} else if (score >= 75) {
    console.log("B");  // prints this
} else if (score >= 60) {
    console.log("C");
} else {
    console.log("F");
}


// =========================================================
// SWITCH — best when comparing one value against many constants
// =========================================================

const day = "Monday";

switch (day) {
    case "Saturday":
    case "Sunday":
        console.log("Weekend");
        break;
    case "Monday":
        console.log("Start of work week"); // prints this
        break;
    case "Friday":
        console.log("Almost weekend");
        break;
    default:
        console.log("Midweek");
}
// break is required — without it, execution "falls through" to the next case


// =========================================================
// WHILE — run while a condition is true
// =========================================================

let count = 0;
while (count < 3) {
    console.log(`while: count = ${count}`);
    count++;
}

// do-while — body always runs at least once
let attempt = 0;
do {
    console.log(`attempt: ${attempt}`);
    attempt++;
} while (attempt < 0); // condition is false immediately, but body ran once


// =========================================================
// FOR — classic counter loop
// =========================================================

for (let i = 0; i < 4; i++) {
    if (i === 2) continue;  // skip 2
    if (i === 4) break;     // never reached here but shows syntax
    console.log(`for: i = ${i}`);  // 0, 1, 3
}


// =========================================================
// FOR...OF — iterate values of an iterable (array, string, Map, Set)
// =========================================================

const fruits = ["apple", "banana", "cherry"];
for (const fruit of fruits) {
    console.log(fruit);
}

// Works on strings too — iterates over characters
for (const char of "hi") {
    console.log(char);  // "h", "i"
}


// =========================================================
// FOR...IN — iterate keys of an object (not recommended for arrays)
// =========================================================

const person = { name: "Alice", age: 30, city: "Warsaw" };
for (const key in person) {
    console.log(`${key}: ${person[key]}`);
}


// =========================================================
// LABELED BREAK — break out of nested loops
// =========================================================

outer: for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
        if (i === 1 && j === 1) break outer;  // exits both loops
        console.log(`(${i}, ${j})`);
    }
}
// Prints: (0,0) (0,1) (0,2) (1,0)  — stops at (1,1)
